### PR TITLE
Registry documentation

### PR DESCRIPTION
**Pull request by nachocano**
Fixes https://github.com/knative/docs/issues/1285

## Proposed Changes

- Adding initial documentation for Event Registry feature

/label cherrypick-0.6

### CODE REVIEW

**Comment by samodell** [comment]
Thanks @nachocano ! I'm taking a look at this now.

File: docs/eventing/event_registry.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+# Event Registry
+
```
Comment (samodell) [inline docs/eventing/event_registry.md]: Update the title of this file to be "event-registry.md" ( changing the underscore to a hyphen) for consistency with the other files in the docs directory.

File: docs/eventing/README.md [hunk 2]
```diff
@@ -53,6 +53,16 @@ Addressable. You can create as many Triggers as necessary.
 
 ![Broker Trigger Diagram](./images/broker-trigger-overview.svg)
 
+### Event registry
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/README.md]: Done!

File: docs/eventing/README.md [hunk 3]
```diff
@@ -53,6 +53,16 @@ Addressable. You can create as many Triggers as necessary.
...
+
+As of v0.6, Knative Eventing defines an EventType object to make it easier for 
+consumers to discover the type of events they can consume from the different
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/README.md]: Done! Thanks!

File: docs/eventing/README.md [hunk 4]
```diff
@@ -53,6 +53,16 @@ Addressable. You can create as many Triggers as necessary.
...
 
+### Event registry
+
+As of v0.6, Knative Eventing defines an EventType object to make it easier for 
+consumers to discover the type of events they can consume from the different 
+Brokers.
+
+The registry consists of a collection of event types. The event types stored in 
+the registry contain (all) the required information for a consumer to create a 
+Trigger without resorting to some other OOB mechanism.
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/README.md]: This one is key to the concept of Registry. It allow to just be able to create a trigger without looking outside the cluster. 
I changed OOB for out-of-band.... If you still not agree, we can further revisit.

File: docs/eventing/event_registry.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
+# Event Registry
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Thanks! Done

File: docs/eventing/event_registry.md (hunk 1 above)
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+## Overview
+
+The Event Registry is a component that maintains a catalog of the event types that can flow through the system.
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Removed, thanks!

File: docs/eventing/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (samodell) [reply docs/eventing/README.md]: ah, okay, that makes sense. Yep, spelling it out definitely helps!

File: docs/eventing/event_registry.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
+# Event Registry
+
+## Overview
+
+The Event Registry is a component that maintains a catalog of the event types that can flow through the system. 
+For doing so, it introduces a new [EventType](../reference/eventing/eventing.md) CRD to be able to persist the event types
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done thanks!

File: docs/eventing/event_registry.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+## Overview
+
+The Event Registry is a component that maintains a catalog of the event types that can flow through the system. 
+For doing so, it introduces a new [EventType](../reference/eventing/eventing.md) CRD to be able to persist the event types 
+information in Kubernetes data store. 
+
+## Before you begin
+
+1. Read about the [Broker and Trigger objects](./broker-trigger.md).
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: It's this [PR](https://github.com/knative/docs/pull/1279) that @grantr created. We need to refer that one here as it has a lot of background information.

File: docs/eventing/event_registry.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+1. Be familiar with the [User stories and personas for Knative eventing](https://docs.google.com/document/d/15uhyqQvaomxRX2u8s0i6CNhA86BQTNztkdsLUnPmvv4/edit?usp=sharing).
+1. Be familiar with the [Sources](./sources/README.md).
+
+## Discovering Events with the Registry
+
+By leveraging the Registry, Event Consumers can discover what are the different types of events that they can consume 
+from the Brokers' event meshes. Our current implementation mainly targets the Broker/Trigger model, and aims to help 
+consumers creating Triggers. 
+
+Event Consumers can simply execute the following command to see what are the events they can *subscribe* to:
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+## Before you begin
+
+1. Read about the [Broker and Trigger objects](./broker-trigger.md).
+1. Be familiar with the [CloudEvents spec](https://github.com/cloudevents/spec/blob/master/spec.md),
+   particularly the [Context Attributes](https://github.com/cloudevents/spec/blob/master/spec.md#context-attributes)
+   section.
+1. Be familiar with the [User stories and personas for Knative eventing](https://docs.google.com/document/d/15uhyqQvaomxRX2u8s0i6CNhA86BQTNztkdsLUnPmvv4/edit?usp=sharing).
+1. Be familiar with the [Sources](./sources/README.md).
+
+## Discovering Events with the Registry
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done!

File: docs/eventing/event_registry.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+Event Consumers can (and in most cases would) create Triggers filtering on this attribute.
+
+- `schema`: is a valid URI with the EventType schema. It may be a JSON schema, a protobuf schema, etc. It is optional.
+
+- `description`: is a string describing what the EventType is about. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. It is mandatory.
+
+
+## Subscribing to Events of Interest
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done!

File: docs/eventing/event_registry.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+## Populating the Registry
+
+Now that we know how to discover events using the Registry and how we can leverage that information to subscribe to 
+events of interest, let's move on to the next topic: How do we actually populate the Registry in the first place? 
+
+You might be wondering why didn't we explain this first? The simple answer is we could have, but as population of the 
+Registry is more of a Cluster Configurator concern rather than an Event Consumer one, we decided to leave it for the end.  
+
+1. Manual Registration
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+For doing so, it introduces a new [EventType](../reference/eventing/eventing.md) CRD to be able to persist the event types 
+information in Kubernetes data store. 
+
+## Before you begin
+
+1. Read about the [Broker and Trigger objects](./broker-trigger.md).
+1. Be familiar with the [CloudEvents spec](https://github.com/cloudevents/spec/blob/master/spec.md),
+   particularly the [Context Attributes](https://github.com/cloudevents/spec/blob/master/spec.md#context-attributes)
+   section.
+1. Be familiar with the [User stories and personas for Knative eventing](https://docs.google.com/document/d/15uhyqQvaomxRX2u8s0i6CNhA86BQTNztkdsLUnPmvv4/edit?usp=sharing).
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+1. Be familiar with the [Sources](./sources/README.md).
+
+## Discovering Events with the Registry
+
+By leveraging the Registry, Event Consumers can discover what are the different types of events that they can consume 
+from the Brokers' event meshes. Our current implementation mainly targets the Broker/Trigger model, and aims to help
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+1. Read about the [Broker and Trigger objects](./broker-trigger.md).
+1. Be familiar with the [CloudEvents spec](https://github.com/cloudevents/spec/blob/master/spec.md),
+   particularly the [Context Attributes](https://github.com/cloudevents/spec/blob/master/spec.md#context-attributes)
+   section.
+1. Be familiar with the [User stories and personas for Knative eventing](https://docs.google.com/document/d/15uhyqQvaomxRX2u8s0i6CNhA86BQTNztkdsLUnPmvv4/edit?usp=sharing).
+1. Be familiar with the [Sources](./sources/README.md).
+
+## Discovering Events with the Registry
+
+By leveraging the Registry, Event Consumers can discover what are the different types of events that they can consume
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done! thanks!

File: docs/eventing/event_registry.md [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+Registry is more of a Cluster Configurator concern rather than an Event Consumer one, we decided to leave it for the end.  
+
+1. Manual Registration
+
+    In order to populate the Registry, a Cluster Configurator can manually register the EventTypes. 
+    This means that the configurator can simply apply EventTypes yaml files, just as with any other Kubernetes resource:
+
+    `kubectl apply -f <event_type.yaml>`
+
+1. Automatic Registration
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done!

File: docs/eventing/event_registry.md [hunk 17]
```diff
@@ -0,0 +1,260 @@
+# Event Registry
+
+## Overview
+
+The Event Registry is a component that maintains a catalog of the event types that can flow through the system. 
+For doing so, it introduces a new [EventType](../reference/eventing/eventing.md) CRD to be able to persist the event types 
+information in Kubernetes data store.
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: changed it to cluster's data store. Let me know if it doesn't work

File: docs/eventing/event_registry.md [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+   particularly the [Context Attributes](https://github.com/cloudevents/spec/blob/master/spec.md#context-attributes)
+   section.
+1. Be familiar with the [User stories and personas for Knative eventing](https://docs.google.com/document/d/15uhyqQvaomxRX2u8s0i6CNhA86BQTNztkdsLUnPmvv4/edit?usp=sharing).
+1. Be familiar with the [Sources](./sources/README.md).
+
+## Discovering Events with the Registry
+
+By leveraging the Registry, Event Consumers can discover what are the different types of events that they can consume 
+from the Brokers' event meshes. Our current implementation mainly targets the Broker/Trigger model, and aims to help 
+consumers creating Triggers.
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+- `schema`: is a valid URI with the EventType schema. It may be a JSON schema, a protobuf schema, etc. It is optional.
+
+- `description`: is a string describing what the EventType is about. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. It is mandatory.
+
+
+## Subscribing to Events of Interest
+
+Given that the consumers now know what events can be consumed from the Brokers' event meshes, they can easily create
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+
+- `description`: is a string describing what the EventType is about. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. It is mandatory.
+
+
+## Subscribing to Events of Interest
+
+Given that the consumers now know what events can be consumed from the Brokers' event meshes, they can easily create 
+Triggers to materialize their desire to subscribe to particular ones.
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+- `description`: is a string describing what the EventType is about. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. It is mandatory.
+
+
+## Subscribing to Events of Interest
+
+Given that the consumers now know what events can be consumed from the Brokers' event meshes, they can easily create 
+Triggers to materialize their desire to subscribe to particular ones. 
+Here are few Trigger examples that do so, using basic exact matching on `type` and/or `source`,
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+
+- `broker` refers to the Broker that can provide the EventType. It is mandatory.
+
+
+## Subscribing to Events of Interest
+
+Given that the consumers now know what events can be consumed from the Brokers' event meshes, they can easily create 
+Triggers to materialize their desire to subscribe to particular ones. 
+Here are few Trigger examples that do so, using basic exact matching on `type` and/or `source`, 
+based on the above Registry output.
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks, also lowercased registry

File: docs/eventing/event_registry.md [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+    
+    Note that events won't be able to be consumed by this Trigger's subscriber until the Broker becomes ready.
+
+
+## Populating the Registry
+
+Now that we know how to discover events using the Registry and how we can leverage that information to subscribe to 
+events of interest, let's move on to the next topic: How do we actually populate the Registry in the first place? 
+
+You might be wondering why didn't we explain this first? The simple answer is we could have, but as population of the
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done thanks!

File: docs/eventing/event_registry.md [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+    If you are interested in more information regarding configuration options of a KafkaSource, please refer to the 
+    [KafKaSource example](https://github.com/knative/eventing-sources/tree/master/contrib/kafka/samples).
+    
+    For this discussion, the relevant information from the yaml above are the `sink` and the `topics`. 
+    We observe that the `sink` is of kind `Broker`. We currently only support automatic creation of EventTypes for Sources 
+    instances that point to Brokers.
+    Regarding `topics`, this is what we use for the EventTypes `source` field, which is equal to the CloudEvent source 
+    attribute. 
+
+    When a Cluster Configurator `kubectl apply` this yaml, not only the KafkaSource `kafka-source-sample` will be instantiated,
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

File: docs/eventing/event_registry.md [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+    but also two EventTypes will be added to the Registry (as there are two topics). You can see that in the Registry
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

**Comment by nachocano** [comment]
Thanks for all the comments @samodell . They were very useful!
Is ready for a second pass

File: docs/eventing/event_registry.md [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+    We observe that the `sink` is of kind `Broker`. We currently only support automatic creation of EventTypes for Sources 
+    instances that point to Brokers.
+    Regarding `topics`, this is what we use for the EventTypes `source` field, which is equal to the CloudEvent source 
+    attribute. 
+
+    When a Cluster Configurator `kubectl apply` this yaml, not only the KafkaSource `kafka-source-sample` will be instantiated,
+    but also two EventTypes will be added to the Registry (as there are two topics). You can see that in the Registry 
+    example output from the previous sections.  
+        
+## Next Steps
```
[the remark this answers is not in the document]
Comment (nachocano) [reply docs/eventing/event_registry.md]: Done, thanks!

**Comment by samodell** [comment]
/lgtm
/approve

Thanks Nacho, this is great!

**Pull request merged**